docs/proposals/2170-kubeflow-training-v2/README.md

```diff
@@ -0,0 +1,1438 @@
+   <td><code>python</code>
+   </td>
+   <td><a href="https://www.tensorflow.org/guide/distributed_training">TensorFlow Distributed</a>
+   </td>
+  </tr>
+  <tr>
+   <td>Jax
+   </td>
+   <td>Process <code>jax.local_devices()</code>
+   </td>
+   <td>Host
+<p>
+<code>jax.devices()</code>
+   </td>
+   <td><code>python</code> or <code>mpirun</code>
+   </td>
+   <td><a href="https://jax.readthedocs.io/en/latest/multi_process.html">Jax Distributed</a>
+   </td>
+  </tr>
+  <tr>
+   <td>PaddlePaddle
+   </td>
+   <td>Worker
+   </td>
+   <td>Node
+   </td>
+   <td><code>python -m paddle.distributed.launch</code>
+   </td>
+   <td><a href="https://www.paddlepaddle.org.cn/documentation/docs/en/guides/06_distributed_training/cluster_quick_start_en.html">Paddle Distributed</a>
+   </td>
+  </tr>
+  <tr>
+   <td>XGBoost
+   </td>
+   <td>Worker
+   </td>
+   <td><em>Not Applicable</em>
+   </td>
+   <td><code>python</code> 
+   </td>
+   <td><a href="https://github.com/dmlc/xgboost/blob/a5a58102e5e82fa508514c34cd8e5f408dcfd3e1/python-package/xgboost/tracker.py#L17">Rabit Tracker</a> for c10d
+   </td>
+  </tr>
+  <tr>
+   <td>DeepSpeed
+   </td>
+   <td>Slot
+   </td>
+   <td>Node
+<p>
+<code>(--num_nodes)</code>
+   </td>
+   <td><code>deepspeed</code>
+   </td>
+   <td><a href="https://www.deepspeed.ai/getting-started/#resource-configuration-multi-node">DeepSpeed Distributed</a>
+   </td>
+  </tr>
+</table>
+
+Additionally, check [this document for the `mpirun` command](https://gist.github.com/vsoch/9ac7c4448dffe656d946edceaa58bd9e)
+for other MPI implementations: Intel MPI, MPICH, Spectrum MPI.
+
+## The TrainJob API
+
+The `TrainJob` exposes APIs that data scientist can override in `TrainingRuntime` to create
+a training job:
+
+```golang
+type TrainJob struct {
+	metav1.TypeMeta   `json:",inline"`
+	metav1.ObjectMeta `json:"metadata,omitempty"`
+
+	// Spec defines the desired state of TrainJob.
+	Spec TrainJobSpec `json:"spec"`
+
+	// Status defines the current state of TrainJob.
+	Status TrainJobStatus `json:"status,omitempty"`
+}
+
+type TrainJobSpec struct {
+	// Reference to the Training Runtime.
+	TrainingRuntimeRef *TrainingRuntimeRef `json:"trainingRuntimeRef"`
+
+	// Parameters that data scientists can override
+	Trainer *Trainer `json:"trainer,omitempty"`
+
+	// Configuration for training dataset
+	DatasetConfig *DatasetConfig `json:"datasetConfig,omitempty"`
+
+	// Configuration for the pre-trained model and location for model output
+	ModelConfig *ModelConfig `json:"modelConfig,omitempty"`
+
+	// Custom metadata to apply for Job, JobSet, etc.
+	Labels      map[string]string `json:"labels,omitempty"`
+	Annotations map[string]string `json:"annotations,omitempty"`
+}
+
+type TrainingRuntimeRef struct {
+	// Name for the training runtime.
+	Name string `json:"name"`
+	// Namespace for the runtime.
+	// If namespace is set, TrainingRuntime is used. Otherwise, ClusterTrainingRuntime is used.
+	Namespace string `json:"namespace,omitempty"`
+}
```

Comment:
As we discussed in Google Doc and previous meeting, could we add the Group/Version/Kind since this switching mechanism depending on the `namespace` field indicates the implied the specifications. It would be better to define all specifications as APIs.

Reply:
Should we initially just introduce Kind since we don't support the runtime versioning ?
Also, are we going to validate that if `Kind: ClusterTrainingRuntime` is set, then `namespace` parameter should be ignored ?

Reply:
Can we add the Kind and Version? In this version doesn't mean the runtime revision, which indicates the API version like v2alpha1.

```yaml
trainingRuntimeRef:
  name: torch-multi-node
  namespace: tenant-a  # When they use the TrainingRuntime, they need to specify this field.
  kind: TrainingRuntime | ClusterTrainingRuntime
  version: v2alpha1 # This is API version
  revision: v1 # This will be supported in the future after we introduce the runtime revision management mechanism.
```

Reply:
Should we re-use: https://github.com/kubernetes/api/blob/master/core/v1/types.go#L585C6-L585C26 ?
Or we don't need the `group` since it is always `kubeflow.org`?